src/vs/workbench/contrib/terminal/browser/xterm/decorationAddon.ts

```diff
@@ -79,11 +83,65 @@ export class DecorationAddon extends Disposable implements ITerminalAddon {
 				this.refreshLayouts();
 			} else if (e.affectsConfiguration('workbench.colorCustomizations')) {
 				this._refreshStyles(true);
+			} else if (e.affectsConfiguration(TerminalSettingId.ShellIntegrationDecorationsEnabled)) {
+				this._updateDecorationVisibility();
 			}
 		});
 		this._themeService.onDidColorThemeChange(() => this._refreshStyles(true));
+		this._updateDecorationVisibility();
 	}
 
+	private _updateDecorationVisibility(): void {
+		const showDecorations = this._configurationService.getValue(TerminalSettingId.ShellIntegrationDecorationsEnabled);
+		switch (showDecorations) {
+			case 'never':
+				this._disposeOfDecorations();
+				this._updateGutterDecorationVisibility(false);
+				this._overviewRulerDecorations = false;
+				this._decorationsDisabled = true;
+				break;
+			case 'both':
+				this._disposeOfDecorations();
+				this._updateGutterDecorationVisibility(true);
+				this._attachToCommandCapability();
+				this._overviewRulerDecorations = true;
+				this._decorationsDisabled = false;
+				this._attachToCommandCapability();
+				break;
+			case 'gutter':
+				this._disposeOfDecorations();
+				this._updateGutterDecorationVisibility(true);
+				this._overviewRulerDecorations = false;
+				this._decorationsDisabled = false;
+				this._attachToCommandCapability();
+				break;
+			case 'overviewRuler':
+				this._updateGutterDecorationVisibility(false);
+				this._overviewRulerDecorations = true;
+				this._decorationsDisabled = false;
+				this._attachToCommandCapability();
+				break;
+		}
+	}
+
+	private _disposeOfDecorations(): void {
+		this._placeholderDecoration?.dispose();
+		for (const value of this._decorations.values()) {
+			value.decoration.dispose();
+			dispose(value.disposables);
+		}
+	}
+
+	private _updateGutterDecorationVisibility(visible?: boolean): void {
+		const containers = document.querySelectorAll(DecorationSelector.Container);
+		for (const container of containers) {
```

Comment:
We can't hide the decoration container as we will use other decorations as well which should still work when command decorations are disables (eg. contextual commands, find matches). You could add a class to the decoration in `onRender` and hide that though.